### PR TITLE
i18n: add registerLocaleData() method

### PR DESCRIPTION
**Pull request by paulirish**
see [Audits Panel Localization Design](https://docs.google.com/document/d/1jnt3BqKB-4q3AE94UWFA0Gqspx8Sd_jivlB7gQMlmfk/edit)

This method enables us to "install" new locale strings into our i18n string lookup.

Unrelated, but I added a few more tests to getFormatted as a driveby.

### CODE REVIEW

File: lighthouse-core/test/lib/i18n/i18n-test.js [hunk 1]
```diff
@@ -98,6 +113,44 @@ describe('i18n', () => {
...
+      const localeData = {
+        'lighthouse-core/audits/is-on-https.js | title': {
+          'message': 'es-419 uses https!'
+        }
+      };
+      i18n.registerLocaleData('es-419', localeData);
+
+      // And confirm that's what is returned
+      const formattedStr2 = i18n.getFormatted(str_(UIStrings.title), 'es-419');
+      expect(formattedStr2).toEqual('es-419 uses https!');
```
Comment (patrickhulce) [inline lighthouse-core/test/lib/i18n/i18n-test.js]: one more assertion that some other string that used to be there is removed? would be nice to clarify in the tests that it's not merging or something it's just totally replacing

File: lighthouse-core/lib/i18n/i18n.js [hunk 2]
```diff
@@ -473,6 +473,17 @@ function replaceIcuMessageInstanceIds(inputObject, locale) {
   return icuMessagePaths;
 }
 
+/** @typedef {import('./locales').LhlMessages} LhlMessages */
+
+/**
+ * Populate the i18n string lookup dict with locale data
```
Comment (patrickhulce) [inline lighthouse-core/lib/i18n/i18n.js]: should we link off to that design and/or mention how this is expected to be used?

File: lighthouse-core/test/lib/i18n/i18n-test.js [hunk 3]
```diff
@@ -98,6 +113,44 @@ describe('i18n', () => {
...
+
+      const UIStrings = {testString: 'en-US string!'};
+      const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+      const formattedStr = i18n.getFormatted(str_(UIStrings.testString), 'en-XZ');
+      expect(formattedStr).toEqual('en-XZ cuerda!');
+    });
+
+    it('overwrites existing locale strings', () => {
+      const filename = 'lighthouse-core/audits/is-on-https.js';
+      const UIStrings = require('../../../../' + filename).UIStrings;
```
Comment (patrickhulce) [inline lighthouse-core/test/lib/i18n/i18n-test.js]: any reason this is dynamic? seems like a static string would be easier to refactor when file is moved, etc.

File: lighthouse-core/lib/i18n/i18n.js [hunk 4]
```diff
@@ -473,6 +473,19 @@ function replaceIcuMessageInstanceIds(inputObject, locale) {
...
+ * Used when the host environment selects the locale and serves lighthouse the intended locale file
+ * @see https://docs.google.com/document/d/1jnt3BqKB-4q3AE94UWFA0Gqspx8Sd_jivlB7gQMlmfk/edit
+ * @param {LH.Locale} locale
```
Comment (brendankenny) [inline lighthouse-core/lib/i18n/i18n.js]: this threatens the little `LH.Locale` ~house of cards~ types. We should either limit to overriding existing locales (verify that `locale` here is actually of type `LH.Locale`) or get rid of `LH.Locale` and type all our locale names as `string`. There might be some side effects of the second option, though, if we depend on that behaving well.

File: lighthouse-core/lib/i18n/i18n.js [hunk 5]
```diff
@@ -473,6 +473,19 @@ function replaceIcuMessageInstanceIds(inputObject, locale) {
...
+ * @param {LhlMessages} lhlMessages
```
Comment (brendankenny) [inline lighthouse-core/lib/i18n/i18n.js]: similarly we should shape check this (a la plugins/budgets.json). All our good error handling code assumes the type is correct (and isn't such good error handling code if it isn't)

File: lighthouse-core/lib/i18n/i18n.js [hunk 6]
```diff
@@ -473,6 +473,19 @@ function replaceIcuMessageInstanceIds(inputObject, locale) {
...
+ */
+function registerLocaleData(locale, lhlMessages) {
+  LOCALES[locale] = lhlMessages;
```
Comment (brendankenny) [inline lighthouse-core/lib/i18n/i18n.js]: I'm a little afraid to make `LOCALES` a mutable singleton like this now, but it's hard to think through the consequences and if they matter. #9296 and #9043 (and wherever the `field-performance` plugin feedback is) point at the fact that people likely will use this outside of devtools, though.

File: lighthouse-core/test/lib/i18n/i18n-test.js [hunk 7]
```diff
@@ -98,6 +113,50 @@ describe('i18n', () => {
...
+      const origFailureTitle = i18n.getFormatted(str_(UIStrings.failureTitle), 'es-419');
+      expect(origFailureTitle).toEqual('No usa HTTPS');
+
+      // Now we declare and register the new string...
+      const localeData = {
+        'lighthouse-core/audits/is-on-https.js | title': {
+          'message': 'es-419 uses https!',
+        },
+      };
+      i18n.registerLocaleData('es-419', localeData);
```
Comment (brendankenny) [inline lighthouse-core/test/lib/i18n/i18n-test.js]: this changes `es-419` for all tests run after this

File: lighthouse-core/lib/i18n/i18n.js (hunk 6 above)
Comment (brendankenny) [reply lighthouse-core/lib/i18n/i18n.js]: > I'm a little afraid to make `LOCALES` a mutable singleton like this now, but it's hard to think through the consequences and if they matter. #9296 and #9043 (and wherever the `field-performance` plugin feedback is) point at the fact that people likely will use this outside of devtools, though.

if we don't want to expose this to anyone but devtools (I believe @exterkamp also preferred not to do that), another option (sorry I didn't think of until right now) is something like 
https://github.com/GoogleChrome/lighthouse/blob/b2dfb06221249d86dab667c241303b53c66203ed/clients/devtools-report-assets.js
basically rewriting `locales.js` to instead access the assets in the module. We could be clever instead of having to do it manually

Comment (brendankenny) [reply lighthouse-core/lib/i18n/i18n.js]: > We could be clever instead of having to do it manually

@connorjclark could even get to use a proxy to rewrite any getter on the `locales.js` object into a request for the resource :)

File: lighthouse-core/lib/i18n/i18n.js (hunk 4 above)
Comment (paulirish) [reply lighthouse-core/lib/i18n/i18n.js]: >  We should either limit to overriding existing locales

in reality this will only be used to override existing locales

File: lighthouse-core/lib/i18n/i18n.js [hunk 8]
```diff
@@ -473,6 +473,19 @@ function replaceIcuMessageInstanceIds(inputObject, locale) {
   return icuMessagePaths;
 }
 
+/** @typedef {import('./locales').LhlMessages} LhlMessages */
+
+/**
+ * Populate the i18n string lookup dict with locale data
+ * Used when the host environment selects the locale and serves lighthouse the intended locale file
+ * @see https://docs.google.com/document/d/1jnt3BqKB-4q3AE94UWFA0Gqspx8Sd_jivlB7gQMlmfk/edit
```
[the remark this answers is not in the document]
Comment (paulirish) [reply lighthouse-core/lib/i18n/i18n.js]: whatcha mean?  it's exposed at the module level..

[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/lib/i18n/i18n.js]: > it's exposed at the module level..

The doc strings, I mean. They aren't in the module-level file and tools like vscode don't forward the jsdocs

File: lighthouse-core/lib/i18n/i18n.js (hunk 4 above)
Comment (brendankenny) [reply lighthouse-core/lib/i18n/i18n.js]: > in reality this will only be used to override existing locales

This is partly what https://github.com/GoogleChrome/lighthouse/pull/9638#discussion_r321013473 is touching on. I don't think we can treat it as a private api
